### PR TITLE
String Case Enhancement

### PR DESCRIPTION
**Pull request by Wealthyturtle**
### Description

- [x] Better Pattern
- [x] Support Lists
- [x] Lenient & Strict Proper Case
- [x] Lenient & Strict Camel Case
- [x] Lenient & Strict Pascal Case
- [x] Lenient, Lower, & Upper Snake Case
- [x] Lenient, Lower, & Upper Kebab Case

Test Code:
```
command /test:
	trigger:
		set {_x} to "TeSt"
		set {_y::*} to "ThE ", "LaZy!+", "bRoWn", "foX", "jumPeD", "oVeR", "tHE", "WaLL", and ""
		broadcast "*%{_x} in uppercase%"
		broadcast "*%{_y::*} in uppercase%"
		broadcast "*%{_x} in lowercase%"
		broadcast "*%{_y::*} in lowercase%"
		broadcast "*%""Oops!"" in lowercase%"
		broadcast "*%""oops!"" in uppercase%"
		broadcast "*%""hellO i'm steve!"" in proper case%"
		broadcast "*%""hellO i'm steve!"" in strict proper case%"
		broadcast "*%""spAwn neW boSs ()"" in camel case%"
		broadcast "*%""spAwn neW boSs ()"" in strict camel case%"
		broadcast "*%""geneRate ranDom numBer ()"" in pascal case%"
		broadcast "*%""geneRate ranDom numBer ()"" in strict pascal case%"
		broadcast "*%""Hello Player!"" in snake case%"
		broadcast "*%""Hello Player!"" in lower snake case%"
		broadcast "*%""Hello Player!"" in upper snake case%"
		broadcast "*%""What is your name?"" in kebab case%"
		broadcast "*%""What is your name?"" in lower kebab case%"
		broadcast "*%""What is your name?"" in upper kebab case%"
```

Output (Valid):
```
*TEST
*THE , LAZY!+, BROWN, FOX, JUMPED, OVER, THE, WALL and
*test
*the , lazy!+, brown, fox, jumped, over, the, wall and
*oops!
*OOPS!
*HellO I'm Steve!
*Hello I'm Steve!
*spAwnNeWBoSs()
*spawnNewBoss()
*GeneRateRanDomNumBer()
*GenerateRandomNumber()
*Hello_Player!
*hello_player!
*HELLO_PLAYER!
*What-is-your-name?
*what-is-your-name?
*WHAT-IS-YOUR-NAME?
```

---
**Target Minecraft Versions:**     Any
**Requirements:**     None
**Related Issues:**     #1860

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprStringCase.java [hunk 1]
```diff
@@ -41,76 +44,118 @@
 /**
  * @author bensku
  */
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/expressions/ExprStringCase.java]: You can remove this because you changed a lot of things and Skript doesn't use them anymore

File: src/main/java/ch/njol/skript/expressions/ExprStringCase.java [hunk 2]
```diff
@@ -41,76 +44,118 @@
...
-		else
-			str = "";
-		
-		return new String[] {str};
+		String[] strs = expr.getAll(e);
+		for (int i = 0; i < strs.length; i++) {
+			if (strs[i] != null) {
+				switch (type) {
+					case 0: // Basic Case Change 
+						strs[i] = (casemode == 1) ? strs[i].toUpperCase(Locale.ENGLISH) : strs[i].toLowerCase(Locale.ENGLISH);
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/expressions/ExprStringCase.java]: `Locale.ENGLISH` shouldn't be used in such case, because the string is not going to be used for something that requires English like an enum name. For example you would except "isim" to return `İSİM`, not `ISIM` for Turkish.

File: src/main/java/ch/njol/skript/expressions/ExprStringCase.java [hunk 3]
```diff
@@ -41,76 +44,118 @@
...
+				}
+			}
+		}
+		return strs;
 	}
 	
 	@Override
 	public boolean isSingle() {
-		return true;
+		return false;
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/expressions/ExprStringCase.java]: `"text" in upper case` is not a list

```suggestion
		return expr.isSingle();
```

File: src/test/skript/tests/regressions/1860-Extra String Utilities.sk [hunk 4]
```diff
@@ -0,0 +1,16 @@
+test "string cases":
```
Comment (bensku) [inline src/test/skript/tests/regressions/1860-Extra String Utilities.sk]: Don't use a regression test when adding new features. Use syntax test instead. This could be syntaxes/expressions/ExprStringCase.

Comment (Wealthyturtle) [reply src/test/skript/tests/regressions/1860-Extra String Utilities.sk]: Done!

File: src/main/java/ch/njol/skript/expressions/ExprStringCase.java [hunk 5]
```diff
@@ -120,12 +160,58 @@ public boolean isSingle() {
...
+		for (int i = 0; i < words.length; i++)
+			buf += strict ? WordUtils.capitalizeFully(words[i]) : WordUtils.capitalize(words[i]);
+		return buf;
+	}
+	
+	@SuppressWarnings("null")
+	private static String toSnakeCase(String str, int mode) {
+		if (mode == 0)
+			return str.replace(' ', '_');
+		char[] arr = str.toCharArray();
```
Comment (bensku) [inline src/main/java/ch/njol/skript/expressions/ExprStringCase.java]: This doesn't support Unicode codepoints above 65536. Please iterate over codepoints instead.

File: src/main/java/ch/njol/skript/expressions/ExprStringCase.java [hunk 6]
```diff
@@ -120,12 +160,58 @@ public boolean isSingle() {
...
+		for (int i = 0; i < arr.length; i++)
+			arr[i] = (arr[i] == ' ') ? '_' : ((mode == 1) ? Character.toUpperCase(arr[i]) : Character.toLowerCase(arr[i]));
+		return String.copyValueOf(arr);
+	}
+	
+	@SuppressWarnings("null")
+	private static String toKebabCase(String str, int mode) {
+		if (mode == 0)
+			return str.replace(' ', '-');
+		char[] arr = str.toCharArray();
```
Comment (bensku) [inline src/main/java/ch/njol/skript/expressions/ExprStringCase.java]: Same Unicode support issue here.

File: src/test/skript/tests/syntaxes/expressions/1860-ExprStringCase.sk [hunk 7]
```diff
@@ -0,0 +1,16 @@
+test "string cases":
```
Comment (bensku) [inline src/test/skript/tests/syntaxes/expressions/1860-ExprStringCase.sk]: Add some weird characters here. Emojis, for example, *should* stay same when your try to capitalize them. This also helps to test the Unicode support that I'd like to see in this expression.
